torchrl/data/tensor_specs.py

```diff
@@ -981,6 +989,90 @@ def __eq__(self, other):
     def to_numpy(self, val: TensorDict, safe: bool = True) -> dict:
         return super().to_numpy(val, safe)
 
+    def to_onehot(self) -> OneHotDiscreteTensorSpec:
+        return OneHotDiscreteTensorSpec(self.space.n, self.device, self.dtype)
+
+
+@dataclass(repr=False)
+class MultiDiscreteTensorSpec(DiscreteTensorSpec):
+    """A concatenation of discrete tensor spec.
+
+    Args:
+        nvec (iterable of integers): cardinality of each of the elements of
+            the tensor.
+        device (str, int or torch.device, optional): device of
+            the tensors.
+        dtype (str or torch.dtype, optional): dtype of the tensors.
+
+    Examples:
+        >>> ts = MultiDiscreteTensorSpec((3,2,3))
+        >>> ts.is_in(torch.tensor([2, 0, 1]))
+        True
+        >>> ts.is_in(torch.tensor([2, 2, 1]))
+        False
+    """
+
+    def __init__(
+        self,
+        nvec: Sequence[int],
+        device: Optional[DEVICE_TYPING] = None,
+        dtype: Optional[Union[str, torch.dtype]] = torch.long,
+    ):
+        dtype, device = _default_dtype_and_device(dtype, device)
+        self._size = len(nvec)
+        shape = torch.Size([self._size])
+        space = BoxList([DiscreteBox(n) for n in nvec])
+        super(DiscreteTensorSpec, self).__init__(
+            shape, space, device, dtype, domain="discrete"
+        )
+
+    def rand(self, shape: Optional[torch.Size] = None) -> torch.Tensor:
+        if shape is None:
+            shape = torch.Size([])
+        x = torch.cat(
+            [
+                torch.randint(
+                    0,
+                    space.n,
+                    torch.Size([1, *shape]),
+                    device=self.device,
+                    dtype=self.dtype,
+                )
+                for space in self.space
+            ]
+        ).squeeze()
+        _size = [self._size] if self._size > 1 else []
+        return x.T.reshape([*shape, *_size])
```

Comment:
This leads to the following warning if the number of dims of `x` is greater than 2
```
<string>:3: UserWarning: The use of `x.T` on tensors of dimension other than 2 to reverse their shape is deprecated and it will throw an error in a future release. Consider `x.mT` to transpose batches of matricesor `x.permute(*torch.arange(x.ndim - 1, -1, -1))` to reverse the dimensions of a tensor. (Triggered internally at  /Users/distiller/project/conda/conda-bld/pytorch_1646756029501/work/aten/src/ATen/native/TensorShape.cpp:2318.)
```